File: xyz.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
)

/**
 * Execution: Go compiles to single binary. Lightweight goroutines for concurrency.
 * Escape analysis determines Stack vs Heap allocation at compile time.
 */

// Function declarations

// Constants: Evaluated strictly at compile time. No dynamic assignment.
// Untyped constants (like Pi below) maintain arbitrary precision until assigned.
const Pi = 3.14159265358979323846
const TypedMaxUsers int = 100

// Entry point
func main() {
	fmt.Println("Hello World")

	/*
		================================
		1. BASICS OF VARIABLES: DEFINITION & CONSTANTS
		=================================
		Syntax options:
		1. var name type = value (Standard, explicit type)
		2. var name = value      (Type inference at compile time)
		3. name := value         (Short variable declaration - idiomatic for local scopes)
	*/
	var explicitVar int = 10
	var inferredVar = 20
	shortVar := 30

	fmt.Printf("Explicit variable: %d\n", explicitVar)
	fmt.Printf("Inferred variable: %d\n", inferredVar)
	fmt.Printf("Short variable: %d\n", shortVar)
	fmt.Printf("Const PiValue: %.5f\n", Pi)
	fmt.Printf("Const MaxBuffer: %d\n\n", TypedMaxUsers)
	// ==========================================
	// 2. BUILT-IN PRIMITIVES
	// ==========================================
	// Go maps types directly to hardware architectures without wrapper classes.
	// --- Integers ---
	// 'int' size is architecture-dependent (64-bit strictly on your Linux/WSL setup).
	var varInt int = 2147483647
	var varInt8 int8 = 127                   // 1 byte: -128 to 127
	var varInt16 int16 = 32767               // 2 bytes
	var varInt32 int32 = 2147483647          // 4 bytes
	var varInt64 int64 = 9223372036854775807 // 8 bytes

	// Unsigned integers (positive only ranges)
	var varUint8 uint8 = 255 // 'byte' is a built-in alias for uint8
	var varByte byte = 255   // Idiomatic for raw binary/ASCII data

	// --- Floating-Point (Decimals) ---
	// IEEE-754 standard. Go requires explicit bit sizes. No generic "float".
	var varFloat32 float32 = 3.1415927       // 4 bytes. ~7 decimal digits of precision
	var varFloat64 float64 = 3.1415926535897 // 8 bytes. ~15 decimal digits. Default choice.

	// --- Character (Rune/Rune for unicode) ---
	// Go does not have a 'char' primitive. It uses 'rune' (alias for int32).
	// Represents a Unicode code point to natively support multi-byte characters.
	var varRune rune = 'A' // Stores numeric 65. Requires single quotes.

	// --- Boolean ---
	var varBool bool = true

	fmt.Printf("int: %d\n", varInt)
	fmt.Printf("int8: %d\n", varInt8)
	fmt.Printf("int16: %d\n", varInt16)
	fmt.Printf("int32: %d\n", varInt32)
	fmt.Printf("int64: %d\n", varInt64)
	fmt.Printf("uint8: %d\n", varUint8)
	fmt.Printf("byte: %d\n", varByte)
	fmt.Printf("float32: %f\n", varFloat32)
	fmt.Printf("float64: %.15f\n", varFloat64)
	fmt.Printf("rune (char): %c (%d)\n", varRune, varRune)
	fmt.Printf("bool: %v\n\n", varBool)

	// ==========================================
	// 3. REFERENCE TYPES & MEMORY
	// ==========================================
	// Go determines Stack vs. Heap allocation via "Escape Analysis" at compile time.
	// If a variable's reference outlives its scope, it "escapes" to the Heap.

	// --- Pointers ---
	// Variables holding the raw memory address of another variable.
	var varInt_val int = 42
	var ptrInt *int = &varInt_val // '&' operator extracts the memory address
	fmt.Printf("Variable: %d\n", varInt_val)
	fmt.Printf("Pointer address: %p\n", ptrInt)
	fmt.Printf("Dereferenced pointer: %d\n", *ptrInt)

	// --- Strings ---
	// Primitive, immutable byte slices. Under the hood, they are a struct (StringHeader)
	// containing a pointer to a read-only backing array and an integer length.
	var varStr string = "Hello"
	fmt.Printf("String: %s\n", varStr)
	fmt.Printf("String length: %d\n", len(varStr))

	// --- Arrays ---
	// Fixed-size, statically allocated. In Go, Arrays are VALUES, not references.
	// Passing an array to a function copies the entire contiguous memory block.
	var varArrInt [4]int = [4]int{10, 20, 30, 40}
	fmt.Printf("Array: %v\n", varArrInt)
	fmt.Printf("Array[0]: %d\n", varArrInt[0])

	// --- Slices ---
	// Dynamic, reference-like arrays. Internally a struct (SliceHeader):
	// { Pointer to backing Array, Length, Capacity }.
	var varSlice []int = []int{10, 20, 30}
	varSlice = append(varSlice, 40, 50)
	fmt.Printf("Slice: %v (length: %d, capacity: %d)\n", varSlice, len(varSlice), cap(varSlice))
	emptySlice := make([]string, 3) // make() allocates memory & initializes the underlying array

	// --- Maps & Channels ---
	// Complex reference types. Must be initialized with make().
	varMap := make(map[string]int) // Hash table
	varChan := make(chan int)      // Concurrency synchronization primitive
	// Maps (hash tables)
	varMap["Alice"] = 30
	varMap["Bob"] = 25
	fmt.Printf("Map: %v\n", varMap)

	// --- Special Type: nil (zero-value for reference types) ---
	// Zero-value for pointers, interfaces, maps, slices, channels, and functions.
	var nilSlice []int = nil // Primitives (int, bool, float) CANNOT be nil.
	fmt.Printf("Nil slice: %v (is nil: %v)\n\n", nilSlice, nilSlice == nil)
	/* ==========================================================================
	   4. ALL OPERATORS
	   ==========================================================================
	*/
	// Relational: ==, !=, >, <, >=, <=
	a, b := 10, 3
	fmt.Printf("\nComparison: %d > %d is %v\n", a, b, a > b)
	fmt.Printf("Comparison: %d < %d is %v\n", a, b, a < b)
	fmt.Printf("Comparison: %d == %d is %v\n", a, b, a == b)

	// Logical: && (AND), || (OR), ! (NOT). Short-circuiting applies.
	fmt.Printf("\nLogical: true && false = %v\n", true && false)
	fmt.Printf("Logical: true || false = %v\n", true || false)
	fmt.Printf("Logical: !true = %v\n", !true)

	// Arithmetic: +, -, *, /, % (Modulo strictly for integers)
	fmt.Printf("Arithmetic: %d + %d = %d\n", a, b, a+b)
	fmt.Printf("Arithmetic: %d - %d = %d\n", a, b, a-b)
	fmt.Printf("Arithmetic: %d * %d = %d\n", a, b, a*b)
	fmt.Printf("Arithmetic: %d / %d = %d\n", a, b, a/b)
	fmt.Printf("Arithmetic: %d %% %d = %d\n", a, b, a%b)

	// Bitwise: & (AND), | (OR), ^ (XOR), << (Left Shift), >> (Right Shift), &^ (Bit Clear)
	// Assignment: =, +=, -=, *=, /=, %=, &=, |=, ^=

	// Unary: ++, --
	// CRITICAL: In Go, ++ and -- are STATEMENTS, not expressions.
	// in C , c++:  ++ is an expression, so it produces a value but in GO ++ is a statement, not an expression.
	// so in c - it says increase i and give me a value
	// but in go it says just increase i.Go doesn’t let i++ produce a value that can be assigned or used in larger expressions.
	// `val = i++` is illegal. There are no prefix variants (++i).
	counter := 5
	fmt.Printf("\nUnary: counter = %d\n", counter)
	counter++
	fmt.Printf("After counter++: %d\n\n", counter)

	/* ==========================================================================
	   5. ALL CONDITIONALS OF IF FAMILY
	   ==========================================================================
	*/
	// Parentheses around conditions are forbidden; braces are mandatory.
	if varBool {
		// Executes if condition is true
	} else if varInt > 10 {
		// Nested/Chained conditions
	} else {
		// Fallback
	}

	// Init statement in if
	// Idiomatic Go: Initialization Statement within If
	// `val` is scoped purely to this if-else chain. Ideal for error checking.
	if val := 10; val > 5 {
		fmt.Printf("Scoped value: %d\n\n", val)
	}

	// ==========================================
	// 6. SWITCH STATEMENT
	// ==========================================
	// --- Switch ---
	// Evaluates top-to-bottom. Stops at the first match.
	// CRITICAL: NO implicit fall-through (unlike C/Java). `break` is automatic.
	switchVar := 2
	switch switchVar {
	case 1:
		// Logic
	case 2, 3:
		// Comma separated acts as a logical OR.
		// If explicit fall-through is needed, use the `fallthrough` keyword.
	default:
		// Fallback
	}

	// Tagless Switch (Acts as a cleaner chained if-else)
	switch {
	case varInt > 100:
		// Logic
	case varBool == false:
		// Logic
	}

	// --- Ternary Operator ---
	// ARCHITECTURAL NOTE: Go DELIBERATELY omits the ternary operator (condition ? true : false).
	// The designers enforced explicit `if-else` blocks to maximize code readability and consistency.
	var ternaryStatus string
	ternaryAge := 20
	if ternaryAge >= 18 {
		ternaryStatus = "Adult"
	} else {
		ternaryStatus = "Minor"
	}

	/* ==========================================================================
	   7. ALL LOOPS
	   ==========================================================================
	   Go has ONLY ONE looping keyword: `for`.
	*/
	// 1. Standard Initializer; Condition; Post-statement
	for i := 0; i < 5; i++ {
		if i == 2 {
			continue // Skips current iteration
		}
		if i == 4 {
			break // Terminates loop immediately
		}
	}

	// 2. While-loop equivalent (Condition only)
	w := 0
	for w < 3 {
		w++
	}

	// 3. Infinite loop
	for {
		break // Exits immediately to prevent hang
	}

	// 4. For-Range (For-Each equivalent)
	// Iterates over slices, arrays, maps, strings, or channels.
	// Returns index/key and a COPY of the element.
	for index, value := range varArrInt {
		_ = index // Blank identifier `_` ignores unused variables (Go compiler requirement)
		_ = value
	}

	/* ==========================================================================
	   8 & 9. ALL INPUT OUTPUT & FORMATTING
	   ==========================================================================
	*/
	// Basic Output
	fmt.Println("Prints string appended with newline")
	fmt.Print("Prints string without newline")

	// Formatting (printf) verbs:
	// %v  (default value format)
	// %+v (structs with field names)
	// %#v (Go-syntax representation)
	// %T  (Type of the value)
	// %d  (Base 10 integer), %f (float), %s (string), %p (pointer address)
	fmt.Printf("\nFormatted Decimal (2 places): %.2f\n", varFloat64)
	fmt.Printf("Memory Address: %p | Type: %T\n", ptrInt, varInt)

	// Standard Input
	// Method 1: fmt.Scan (Reads space-separated tokens. Leaves '\n' in buffer)
	var inputAge int
	fmt.Print("Enter age: ")
	// fmt.Scan(&inputAge) // Requires pointer memory address to mutate variable

	// Method 2: bufio for full-line reading (Includes spaces)
	reader := bufio.NewReader(os.Stdin)
	fmt.Print("Enter full name: ")
	// nameInput, _ := reader.ReadString('\n')
	// _ -> dont care about the error
	// _, err := reader.ReadString('\n') // ignore string
	// (\n is included; use strings.TrimSpace(name) to remove it.)
	_ = reader
	// means "I'm intentionally not using reader yet." 
	// It prevents the Go compiler from reporting an "unused variable" error.


	// Input examples (commented, for manual testing)
	fmt.Println("\nInput example (commented - uncomment to test):")
	// reader := bufio.NewReader(os.Stdin)
	// fmt.Print("Enter your name: ")
	// name, _ := reader.ReadString('\n')
	// name = strings.TrimSpace(name)
	// fmt.Printf("Hello, %s!\n", name)

	// Demo without actual input
	demoName := "John"
	demoAge := 30
	fmt.Printf("Demo name: %s, age: %d\n\n", demoName, demoAge)

	// ==========================================
	// 9. PARSING & TYPE CONVERSION
	// ==========================================
	// --- Implicit/Explicit Casting ---
	// Go STRICTLY forbids implicit type coercion. You must explicitly cast.
	var rawInt int = 50
	var castFloat float64 = float64(rawInt)

	// --- Parsing Strings to Primitives ---
	strNum := "25"
	// strconv.Atoi (ASCII to Integer). Returns value and an error object.
	parsedAge, err := strconv.Atoi(strNum)
	if err == nil {
		_ = parsedAge
	}

	// String to int
	strNum = "25"
	parsedInt, err := strconv.Atoi(strNum)
	if err == nil {
		fmt.Printf("String '%s' to int: %d\n", strNum, parsedInt)
	}

	// String to float
	strFloat := "3.14"
	parsedFloat, _ := strconv.ParseFloat(strFloat, 64)
	fmt.Printf("String '%s' to float: %.2f\n", strFloat, parsedFloat)

	strFloat = "3.14"
	// 64 specifies IEEE 754 64-bit precision
	parsedFloat, _ = strconv.ParseFloat(strFloat, 64)

	// --- Converting Primitives to Strings ---
	score := 99
	// Method 1: Itoa (Integer to ASCII)
	strScore1 := strconv.Itoa(score)
	// Method 2: Sprintf (String Print Format - formats without outputting)
	strScore2 := fmt.Sprintf("%d", score)

	// ==========================================
	// 10 & 11. FUNCTIONS & PARAMETERS
	// ==========================================
	val1 := 10
	val2 := 10

	// ARCHITECTURAL CORE: In Go, EVERYTHING is pass-by-value.
	// When you pass a pointer, you are passing a COPY of the memory address.

	// Pass by value
	originalVal := 10
	fmt.Printf("\nBefore passByValue: %d\n", originalVal)
	demonstratePassByValue(originalVal) // Stack copy mutated. val1 remains 10.
	fmt.Printf("After passByValue: %d (unchanged)\n", originalVal)

	// Pass by pointer (Go's pass-by-reference)
	originalRef := 10
	fmt.Printf("\nBefore passByPointer: %d\n", originalRef)
	demonstratePassByPointer(&originalRef) // Address copied. Dereferencing mutates original. val2 becomes 99.
	fmt.Printf("After passByPointer: %d (changed!)\n", originalRef)

	// Slice pass (reference semantics)
	originalSlice := []int{1, 2, 3, 4, 5}
	fmt.Printf("\nBefore slicePass: %v\n", originalSlice)
	demonstrateSlicePass(originalSlice)
	fmt.Printf("After slicePass: %v\n", originalSlice)

	// Variadic function
	total := sumAll(1, 2, 3, 4, 5)
	fmt.Printf("\nsumAll(1,2,3,4,5) = %d\n", total)

	// Multiple return values
	result, err := divide(10, 2)
	if err == nil {
		fmt.Printf("divide(10, 2) = %d\n", result)
	}

	// _ = reader // Silence unused warning if not testing input
	_ = emptySlice
	_ = varChan
	_ = ternaryStatus
	_ = inputAge
	_ = rawInt
	_ = castFloat
	_ = strScore1
	_ = strScore2
	_ = val1
	_ = val2

	// Silencing strict compiler "unused variable" errors
	// _, _, _, _, _, _, _, _ = explicitVar, inferredVar, shortVar, varInt8, varInt16, varInt32, varInt64, varUint8
	// _, _, _, _, _, _, _, _ = varByte, varFloat32, varRune, varStr, varSlice, emptySlice, varMap, varChan
	// _, _, _, _, _, _ = nilPtr, ternaryStatus, inputAge, castFloat, parsedFloat, strScore1
	// _, _ = strScore2, err
}

// ==========================================
// FUNCTION IMPLEMENTATIONS
// ==========================================
// Syntax: func funcName(paramName paramType) returnType { ... }
// - Functions are defined at the package level.
// - Go supports Multiple Return Values, predominantly used for Error Handling.

// Example of multiple return values
func divide(a, b int) (int, error) {
	if b == 0 {
		return 0, fmt.Errorf("division by zero")
	}
	return a / b, nil
}

// Pass by Value
// `val` receives an isolated, independent copy of the primitive in this stack frame.
func demonstratePassByValue(val int) {
	val = 999
	fmt.Printf("Inside function (pass by value): %d\n", val)
}

// Pass by Pointer (Go's implementation of Pass by Reference)
// `ptr` receives a copied memory address. The `*` operator dereferences it,
// reaching across stack frames to mutate the original variable.
func demonstratePassByPointer(ptr *int) {
	*ptr = 999
	fmt.Printf("Inside function (pass by pointer): %d\n", *ptr)
}
func demonstrateSlicePass(slice []int) {
	if len(slice) > 0 {
		slice[0] = 999
		fmt.Printf("Inside function (slice modified): %v\n", slice)
	}
}

func addNumbers(a, b int) int {
	return a + b
}

func sumAll(numbers ...int) int {
	total := 0
	for _, num := range numbers {
		total += num
	}
	return total
}

// This line serves only to avoid unused import warning in this demo
var reader *bufio.Reader = bufio.NewReader(os.Stdin)

File: xyz.rs
use std::io::{self, Write};


// Function declarations
fn demonstrate_move(s: String);
fn demonstrate_borrow(s: &String);
fn demonstrate_borrow_mut(s: &mut String);
fn add_numbers(a: i32, b: i32) -> i32;
fn sum_all(numbers: &[i32]) -> i32;

const PI: f64 = 3.14159265358979323846;
const MAX_BUFFER: u32 = 1024;

fn main() {
    println!("Hello World");

    // ==========================================
    // 1. BASIC VARIABLES & CONSTANTS
    // ==========================================
    /*
    Variables in Rust are IMMUTABLE by default for thread safety and memory safety.
    Declared using `let`. 
    */
    let x = 5; // Immutable variable
    // x = 6; // ERROR: cannot assign twice to immutable variable

    let mut y = 10; // MUTABLE variable (explicitly opted-in using `mut`)
    y = 15; 

    /*
    Shadowing:
    You can redeclare a variable with the same name. This destroys the previous 
    variable and creates a new one, allowing type changes without needing `mut`.
    */
    // Shadowing (rebind with potentially different type)
    let shadowed = "text";
    let shadowed = shadowed.len(); // Changed from &str to usize

    /*
    Constants and Stastics:
    `const`: Always immutable. Evaluated at compile time. Inlined where used. Must have explicit type.
    `static`: Similar to const, but has a fixed memory address (lives for the entire program).
    */
    const MAX_SPEED: u32 = 100_000;
    static GLOBAL_COUNTER: i32 = 0;


    // ==========================================
    // 2. PRIMITIVES (Scalar & Compound)
    // ==========================================
    /*
    Primitives in Rust have fixed, known sizes at compile time and are pushed onto the Stack.
    
    Integers:
    Signed (i)   : i8, i16, i32 (default), i64, i128, isize (arch-dependent pointer size)
    Unsigned (u) : u8, u16, u32, u64, u128, usize (used for indexing)
    */
    let var_u8: u8 = 255; 
    let var_i32: i32 = 2_147_483_647; // Underscores for readability
    
    /*
    Floating-Point:
    f32 (single precision), f64 (double precision - default)
    IEEE-754 standard.
    */
    let var_f32: f32 = 3.14f32; // Explicit suffix
    let var_f64 = 3.141592653589793; // Inferred as f64

    /*
    Boolean: 1 byte
    */
    let var_bool: bool = true;

    /*
    Character: 4 bytes
    Represents a Unicode Scalar Value (supports emojis, CJK, etc., not just ASCII). Use single quotes.
    */
    let var_char: char = '🦀';

    /*
    Compound Types: Tuples and Arrays (Stack allocated)
    */
    // Tuple: Fixed length, mixed types.
    let var_tuple: (i32, f64, char) = (500, 6.4, 'A');
    let tuple_val = var_tuple.1; // Access via dot notation

    // Array: Fixed length, single type.
    let var_arr: [i32; 4] = [10, 20, 30, 40];
    let arr_init = [0; 100]; // Array of 100 zeros

    println!("Primitives - u8: {}, i32: {}, f32: {}, f64: {}, bool: {}, char: {}", var_u8, var_i32, var_f32, var_f64, var_bool, var_char);
    println!("Tuple: {:?}, value at index 1: {}", var_tuple, tuple_val);
    println!("Array: {:?}, initialized array length: {}\n", var_arr, arr_init.len());


    /*
    Rust's memory model
    - Ownership: Every value has one owner; memory is freed automatically when the owner goes out of scope.
    - Stack: Fast storage for fixed-size values.
    - Heap: Dynamic storage for values whose size isn't known at compile time (e.g. String, Vec).
    - References (&, &mut): Borrow data without taking ownership.
    */
    // ==========================================
    // 3. REFERENCE TYPES & HEAP ALLOCATION
    // ==========================================

    // stack variable
    // String is Heap-allocated, dynamic, UTF-8 encoded.
    // The Stack holds the pointer, length, and capacity. The Heap holds the characters.
    let mut heap_string = String::from("Hello"); 
    println!("Stack int: {}", x);

    // `&str` (String Slice) is a reference to string data. It is an immutable view.
    // This points to a hardcoded string literal embedded in the binary (static memory).
    let stack_str: &str = "World"; 
    println!("String slice: {}", stack_str);

    // Borrowing  (immutable reference)
    let ref_to_string: &String = &heap_string; // Immutable borrow (&)
    println!("Borrowed reference: {}", ref_to_string);
    let ref_mut_string: &mut String = &mut heap_string; // Mutable borrow (&mut)
    ref_mut_string.push_str(" Rust");  
    println!("After mutable borrow: {}\n", heap_string);

    // Vectors (Dynamic arrays on the Heap)
    let mut var_vec: Vec<i32> = vec![1, 2, 3];
    var_vec.push(4);

    // Box (heap allocation)
    // Smart Pointers (Box allocates directly on the Heap)
    let boxed_int: Box<i32> = Box::new(5);
    println!("Boxed int: {}\n", boxed_int);

    // Null does NOT exist in Rust. It uses the `Option` enum to handle absence of value safely.
    // Option (safe null handling)
    let some_value: Option<i32> = Some(5);
    let no_value: Option<i32> = None;
    println!("Some value: {:?}", some_value);
    println!("No value: {:?}\n", no_value);

    // ==========================================
    // 4. OPERATORS
    // ==========================================
    /*
    Comparison: ==, !=, >, <, >=, <=
    Logical: &&, ||, !
    Arithmetic: +, -, *, /, %
    Assignment: =, +=, -=, *=, /=, %=
    Bitwise: &, |, ^, <<, >>
    
    NOTE: Rust does NOT have increment/decrement (++, --). Use += 1 or -= 1.
    */

    // ==========================================
    // 5. CONDITIONALS (If Family)
    // ==========================================
    /*
    `if` conditions MUST evaluate to a strict `bool` (no implicit truthy/falsy).
    `if` is an EXPRESSION in Rust, meaning it returns a value.
    */
    let condition = true;
    if condition {
        // execute
    } else if 10 > 5 {
        // execute
    } else {
        // execute
    }

    // `if let` (Pattern matching shortcut for one condition)
    let optional_name = Some("Alice");
    if let Some(name) = optional_name {
        println!("Optional name: {}", name);
    }



    // ==========================================
    // 6. MATCH & TERNARY
    // ==========================================

    /*
    Switch equivalent: `match`
    Powerful pattern matching. It MUST be exhaustive (cover all possible values).
    */
    let match_var = 2;
    let match_result = match match_var {
        1 => "One",
        2 | 3 => "Two or Three", // Multiple patterns
        4..=10 => "Four to Ten", // Inclusive range
        _ => "Anything else",    // Default/catch-all (required for exhaustiveness)
    };
    println!("Match result for {}: {}", match_var, match_result);

    /*
    Ternary: Rust has no `? :` operator. Since `if` is an expression, use it directly.
    The return types of all branches must match exactly.
    */
    let ternary_age = 20;
    let ternary_status = if ternary_age >= 18 { "Adult" } else { "Minor" };
    println!("Age {} is: {}\n", ternary_age, ternary_status);

    // ==========================================
    // 7. LOOPS
    // ==========================================
    /*
    loop: Infinite loop. Can return a value.
    */
    let mut counter = 0;
    let loop_result = loop {
        counter += 1;
        if counter == 10 {
            break counter * 2; // Returns 20 from the loop
        }
    };
    println!("Loop result: {}", loop_result);

    /*
    while: Standard condition loop.
    */
    counter = 10;
    while counter > 0 {
        counter -= 1;
    }

    /*
    for: Iterator loop. The standard way to iterate in Rust.
    */
    for i in 0..5 { // Exclusive range: 0, 1, 2, 3, 4
        // execute
    }
    for i in 0..=5 { // Inclusive range: 0 to 5
        // execute
    }
    
    // For-each on collections. iter() borrows, into_iter() moves/consumes.
    for element in var_arr.iter() {
        // execute
    }

    // Loop Labels (break/continue specific outer loops)
    'outer: for x in 0..5 {
        'inner: for y in 0..5 {
            if x * y > 10 {
                break 'outer; // Exits the outermost loop immediately
            }
        }
    }
    // Labeled loop break
    println!("Labeled loop:");
    'outer: for x in 0..5 {
        for y in 0..5 {
            if x * y > 6 {
                println!("Breaking outer at x={}, y={}", x, y);
                break 'outer;
            }
        }
    }
    println!();


    // ==========================================
    // 8. INPUT / OUTPUT
    // ==========================================
    /*
    Standard Output: Handled via macros (`println!`, `print!`)
    */
    println!("println! with newline");
    print!("print! ");
    println!("without leading newline");

    /*
    Standard Input: `std::io::stdin()`
    Reads input as a String. Returns a `Result` type which must be handled (e.g., using `.expect()`).
    */
    let mut input = String::new();
    // io::stdin().read_line(&mut input).expect("Failed to read line");


    // ==========================================
    // 9. FORMATTING
    // ==========================================
    /*
    Macros: `println!` (print to stdout), `format!` (returns a String).
    Requires the type to implement specific Traits:
    `Display` trait ({}): User-facing, pretty print.
    `Debug` trait ({:?}): Developer-facing, shows internal structure. Pretty debug: {:#?}
    */
    // Formatting with format specifiers
    let price = 19.9934;
    let population = 1500000;
    let f_price = 19.9934;
    println!("Population: {}", population);
    println!("Multiple: {} and {}", 42, "text");
    println!("Price: {:.2}", f_price); // 2 decimal places
    println!("Price variable: {:.2}", price);

    let f_num = 5;
    println!("{:<10} {:>5}", "Apple", f_num); // Left-align 10 chars, Right-align 5 chars
    
    println!("Positional: {1}, {0}, {1}", "Zero", "One"); // One, Zero, One
    println!("Named: {first} {second}", first="A", second="B");

    // Debug formatting
    let numbers = vec![1, 2, 3, 4, 5];
    println!("Debug format: {:?}", numbers);
    println!("Pretty debug:\n{:#?}\n", numbers);

    // Manual input example (commented)
    println!("(Input example commented - uncomment to test)");
    // let mut input = String::new();
    // io::stdin().read_line(&mut input).expect("Failed to read line");
    // println!("You entered: {}", input.trim());


    // ==========================================
    // 10. PARSING & TYPE CONVERSION
    // ==========================================
    /*
    Explicit casting (Narrowing/Widening primitives) using `as`.
    */
    let int_val = 10_i32;
    let float_val = int_val as f64;
    println!("Cast i32 to f64: {} -> {}", int_val, float_val);

    /*
    Parsing Strings to Types using `.parse::<T>()`.
    Requires the target type to implement the `FromStr` trait.
    Returns a `Result<T, Error>` because parsing can fail.
    */
    let str_num = "42";
    let parsed_int: u32 = str_num.parse().expect("Not a number!");
    println!("Parse '{}' to u32: {}", str_num, parsed_int);
    // OR using turbofish syntax: string_num.parse::<u32>().unwrap();

    /*
    Converting Types to Strings
    Types that implement `Display` automatically implement `.to_string()`.
    */
    let num_to_string: String = 99.to_string();
    println!("99 to string: \"{}\"", num_to_string);
    let num_to_string_fmt: String = format!("{}", 99);
    println!("Using format!: \"{}\"\n", num_to_string_fmt);


    // ==========================================
    // 11 & 12. FUNCTIONS, PARAMETERS, PASSING
    // ==========================================
    /*
    Functions define parameters with strict types. Return types follow `->`.
    Rust relies on "Expressions vs Statements".
    - Statement: Ends in `;`, performs an action, returns `()` (unit type).
    - Expression: Evaluates to a value. The last expression in a block (without `;`) is automatically returned.
    */

    let sum = add_numbers(5, 10);
    println!("add_numbers(5, 10) = {}", sum);

    // Ownership transfer (move)
    let original_string = String::from("Hello");
    println!("\nBefore demonstrateMove: {}", original_string);
    demonstrate_move(original_string);
    // println!("After demonstrateMove: {}", original_string); // ERROR: original_string was moved!

    // Borrowing (immutable reference)
    let borrowed_string = String::from("World");
    println!("\nBefore demonstrateBorrow: {}", borrowed_string);
    demonstrate_borrow(&borrowed_string);
    println!("After demonstrateBorrow: {} (still available)", borrowed_string);

    // Mutable borrowing
    let mut mutable_string = String::from("Rust");
    println!("\nBefore demonstrateBorrowMut: {}", mutable_string);
    demonstrate_borrow_mut(&mut mutable_string);
    println!("After demonstrateBorrowMut: {}", mutable_string);

    // Slice parameter
    let numbers = vec![10, 20, 30, 40, 50];
    let array_sum = sum_all(&numbers);
    println!("\nsum_all([10,20,30,40,50]) = {}\n", array_sum);


    // Pass by Value (Move or Copy semantics)
    // - Primitives implement the `Copy` trait (bit-for-bit duplication on stack).
    // - Complex types (String, Vec) implement `Move` (ownership is transferred, original variable invalidated).
    fn takes_ownership(s: String) -> String {
        s // Implicit return (no semicolon)
    }

    // Pass by Reference (Borrowing)
    // - `&T`: Immutable reference. Function can read but not alter original data.
    // - `&mut T`: Mutable reference. Function can alter original data in place.
    fn modify_in_place(num: &mut i32, _read_only_str: &str) {
        *num += 10; // Dereference pointer to modify actual value
        // read_only_str cannot be modified here.
    }

    let mut mutable_num = 5;
    modify_in_place(&mut mutable_num, stack_str);
    println!("After modify_in_place: {}", mutable_num);

    // Use all constants and variables
    println!("\nConstants - PI: {}, MAX_BUFFER: {}, MAX_SPEED: {}, GLOBAL_COUNTER: {}", PI, MAX_BUFFER, MAX_SPEED, GLOBAL_COUNTER);
    println!("Shadowed variable final value: {}", shadowed);
    println!("y value: {}", y);
    println!("Input buffer: \"{}\"", input);
    let _ = takes_ownership(String::from("test"));
}


// ==========================================
// FUNCTION IMPLEMENTATIONS
// ==========================================

fn add_numbers(a: i32, b: i32) -> i32 {
    a + b  // Implicit return (no semicolon)
}

// Demonstrates ownership/move semantics
// s is moved into this function, original binding becomes invalid
fn demonstrate_move(s: String) {
    println!("Inside demonstrateMove: {}", s);
    // s is dropped here, memory freed
}

// Demonstrates immutable borrowing
// s is borrowed (referenced), original remains valid
fn demonstrate_borrow(s: &String) {
    println!("Inside demonstrateBorrow: {}", s);
    // s is reference, nothing is dropped
}

// Demonstrates mutable borrowing
// s is mutably borrowed, can modify original
fn demonstrate_borrow_mut(s: &mut String) {
    s.push_str("!");
    println!("Inside demonstrateBorrowMut: {}", s);
}

// Function with slice parameter
// Borrows the data, doesn't take ownership
fn sum_all(numbers: &[i32]) -> i32 {
    let mut total = 0;
    for &num in numbers {
        total += num;
    }
    total
}
